src/android/UploadTask.java

```diff
@@ -0,0 +1,326 @@
+                .putInt(KEY_OUTPUT_STATUS_CODE, response.code());
+
+        // Try read the response body, if any
+        try {
+            final String res = response.body() != null ? response.body().string() : "";
+            final String filename = "upload-response-" + getId() + ".cached-response";
+
+            try (FileOutputStream fos = getApplicationContext().openFileOutput(filename, Context.MODE_PRIVATE)) {
+                fos.write(res.getBytes(StandardCharsets.UTF_8));
+            }
+
+            outputData.putString(KEY_OUTPUT_RESPONSE_FILE, filename);
+
+        } catch (IOException e) {
+            // Should never happen, but if it does it has something to do with reading the response
+            Log.e(TAG, "doWork: Error while reading the response body", e);
+
+            // But recover and replace the body with something else
+            outputData.putString(KEY_OUTPUT_RESPONSE_FILE, null);
+        }
+
+        return Result.success(outputData.build());
+    }
+
+    /**
+     * Called internally by the custom request body provider each time 8kio are written.
+     */
+    private void handleProgress(long bytesWritten, long totalBytes) {
+        if (isStopped()) {
+            currentCall.cancel();
+            return;
+        }
+
+        double percent = ((double) bytesWritten / (double) totalBytes) * 100.0;
+        Log.i(TAG, "handleProgress: " + getId() + " Progress: " + (int) percent);
+
+        setProgressAsync(new Data.Builder()
+                .putString(KEY_PROGRESS_ID, getInputData().getString(KEY_INPUT_ID))
+                .putInt(KEY_PROGRESS_PERCENT, (int) percent)
+                .build()
+        );
+        setForegroundAsync(createForegroundInfo((int) percent));
+    }
+
+    /**
+     * Create the OkHttp request that will be used, already filled with input data.
+     *
+     * @return A ready to use OkHttp request
+     * @throws FileNotFoundException If the file to upload can't be found
+     */
+    @NonNull
+    private Request createRequest() throws FileNotFoundException {
+        final String filepath = getInputData().getString(KEY_INPUT_FILEPATH);
+        assert filepath != null;
+        final String fileKey = getInputData().getString(KEY_INPUT_FILE_KEY);
+        assert fileKey != null;
+
+        // Build URL ...
+        HttpUrl.Builder urlBuilder = Objects.requireNonNull(HttpUrl.parse(getInputData().getString(KEY_INPUT_URL))).newBuilder();
+
+        // ... with parameters if any
+        final int parametersCount = getInputData().getInt(KEY_INPUT_PARAMETERS_COUNT, 0);
+        if (parametersCount > 0) {
+            final String[] parameterNames = getInputData().getStringArray(KEY_INPUT_PARAMETERS_NAMES);
+            assert parameterNames != null;
+
+            for (int i = 0; i < parametersCount; i++) {
+                final String key = parameterNames[i];
+                final Object value = getInputData().getKeyValueMap().get(KEY_INPUT_PARAMETER_VALUE_PREFIX + i);
```

Comment:
C'est pas mieux de définir ces variables avant la boucle ?

Reply:
I see why you would but variables are hoisted anyway so it will result in the exact same perfomances but with wrong scoping.